llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
+    GOHash.update(GODepth - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    DigestType Digest;
+    GOHash.final(Digest);
+    Changed = true;
+    return std::make_tuple(StateIt->second, Digest);
   }
 
-  GOHash.update('T');
-  auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  bool Changed = State.Changed;
+  GOHash.update(static_cast<char>(Tags::GO));
+  auto State = accumulateGODigest(GO, GOHash, HasRepoDefinitionMD);
   // The hashes of the GO's 'contributions' and 'dependencies' are separately
   // added to the GO's hash since their respective arrows point are in different
   // direction. In addition, 'contributions' cannot form loops since 1) only
   // global variables have 'contributions' and 2) all 'contributions' are
   // functions. Therefore, the 'contributions' can be simplified not to record
   // 'visited' table, which avoids having two 'visited' maps. Firstly, add the
   // GO's contributions to its hash.
-  Changed = updateDigestUseContributions(GOHash, State.Contributions, GOIMap) ||
-            Changed;
+  updateDigestUseContributions(GOHash, State.Contributions);
 
   // Add the GO's dependencies to its hash.
-  Changed = updateDigestUseDependencies(GOHash, State.Dependencies, Visited,
-                                        GOIMap, AccumulateGODigest) ||
-            Changed;
-
-  return Changed;
+  const auto result = updateDigestUseDependencies(
+      GO, GOHash, GODepth, State.Dependencies, HasRepoDefinitionMD);
+  Visited.erase(GO);
+  LLVM_DEBUG(dbgs() << "Removing \"" << GO->getName()
+                    << "\" from the Visited map\n");
+  return result;
 }
 
-// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M) {
-  GOInfoMap GOIMap;
-  GONumber GONum;
-
-  for (auto &GV : M.globals()) {
-    if (GV.isDeclaration())
-      continue;
-    calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
-  }
-  for (auto &Fn : M.functions()) {
-    if (Fn.isDeclaration())
-      continue;
-    calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
-  }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+DigestType ModuleHashGenerator::calculateDigest(const GlobalObject *GO,
+                                                bool HasRepoDefinitionMD) {
+  auto const Result = std::get<DigestType>(
+      updateDigestUseDependenciesAndContributions(GO, HasRepoDefinitionMD));
+  assert(Visited.empty() && "Visited is expected to be empty!");
+  return Result;
 }
 
-std::tuple<bool, unsigned, unsigned> generateRepoDefinitions(Module &M) {
-  bool Changed = false;
-  GONumber GONum;
-  GOStateMap Visited;
-  Visited.reserve(M.size() + M.getGlobalList().size());
-  GOInfoMap GOIMap;
+void ModuleHashGenerator::digestModule(Module &M) {
+  auto GOSize = M.size() + M.getGlobalList().size();
+  Visited.reserve(GOSize);
```

Comment:
This doesn't seem right now. This doesn't need to hold the whole thing, just the deepest point.

Reply:
Here, reserve `visited` size to avoid the memory allocation. It will reserve the `possible` deepest path which is `GOSize`.